docs/topics/multiplatform-mobile/multiplatform-mobile-upgrade-app.md

```diff
@@ -0,0 +1,348 @@
+
+1. In `iosApp/iosApp/ContentView.swift`, create a `ViewModel` class for `ContentView`, which will prepare and manage data
+   for it:
+
+    ```Swift
+    import SwiftUI
+    import shared
+    
+    struct ContentView: View {
+        @ObservedObject private(set) var viewModel: ViewModel
+    
+        var body: some View {
+            Text(viewModel.text)
+        }
+    }
+     
+    extension ContentView {
+        class ViewModel: ObservableObject {
+            @Published var text = "Loading..."
+            init() {
+                // Data will be loaded here
+            }
+        }
+    }
+    ```
+   
+    * `ViewModel` is declared as an extension to `ContentView`, as they are closely connected.
+    * The [Combine framework](https://developer.apple.com/documentation/combine) connects the view model (ContentView.ViewModel)
+    with the view (ContentView).
+    * `ContentView.ViewModel` is declared as an `ObservableObject`.
+    * `@Published` wrapper is used for the `text` property.
+    * `@ObservedObject` property wrapper is used to subscribe to the view model.
+
+    Now the view model will emit signals whenever this property changes.
+
+2. Call the `greeting()` function, which now also loads data from the SpaceX API, and save the result in the `text` property:
+
+    ```Swift
+    class ViewModel: ObservableObject {
+        @Published var text = "Loading..."
+            init() {
+                Greeting().greeting { greeting, error in
+                    DispatchQueue.main.async {
+                        if let greeting = greeting {
+                            self.text = greeting
+                        } else {
+                            self.text = error?.localizedDescription ?? "error"
+                        }
+                    }
+                }
+            }
+        }
+    }
+    ```
+  
+   * Kotlin/Native [provides bidirectional interoperability with Objective-C](https://kotlinlang.org/docs/native-objc-interop.html#mappings), thus
+   Kotlin concepts, including `suspend` functions, are mapped to appropriate Swift/Objective-C and vice versa. When you
+   compile a Kotlin module into an Apple framework, suspending functions are available in it as functions with
+   callbacks (`completionHandler`).
+   * The `greeting()` function was marked with the `@Throws(Exception::class)` annotation. So any exceptions that are
+   instances of the `Exception` class or its subclass will be propagated as `NSError`, so you can handle them in the `completionHandler`.
+   * When calling Kotlin `suspend` functions from Swift, completion handlers might be called on threads other than the main,
+   see the [new memory manager migration guide](https://github.com/JetBrains/kotlin/blob/master/kotlin-native/NEW_MM.md#new-memory-manager-migration-guide).
+   That's why `DispatchQueue.main.async` is used to update `text` property, as all UI updates on other threads than main
+   are now allowed in Swift.
+
```

Comment:
We also need to change 'ContentView()' to 'ContentView(viewModel: ContentView.ViewModel())' in iOSApp.swift to make the code compile